joindinapp/Classes/SettingsViewController.m

```diff
@@ -145,21 +180,62 @@ - (void) savePrefs {
 	//[APICaller clearCache];
 }
 
-- (void)gotUserValidateData:(BOOL)success error:(APIError *)err {
-	[self.uiChecking stopAnimating];
+- (void)gotUserValidateData:(BOOL)success error:(APIError *)err data:(NSDictionary *)data {
+	self.uiOk.hidden = NO;
 	if (success) {
-		[self savePrefs];
-		[self.navigationController popViewControllerAnimated:YES];
+		NSMutableDictionary *newParams = [[NSMutableDictionary alloc] initWithDictionary:data];
+		[newParams setObject:self.uiUser.text forKey:@"username"];
+		[self setPrefs:newParams];
+
+		// request the user's details so we can get their gravatar image
+		NSString *verboseURI = [NSString stringWithFormat:@"%@?verbose=yes", [data objectForKey:@"user_uri"]];
+		UserGetDetail *userGetDetail = [APICaller UserGetDetail:self];
+		[userGetDetail call:verboseURI];
+
+		[self toggleSignedIn:YES];
+		[self.uiChecking stopAnimating];
 	} else {
+		[self.uiChecking stopAnimating];
 		UIAlertView *alert;
 		alert = [[UIAlertView alloc] initWithTitle:@"Error" message:@"Invalid username/password"
 										  delegate:self cancelButtonTitle:@"OK" otherButtonTitles:nil];
 		[alert show];
 		[alert release];
-		self.uiOk.hidden = NO;
 	}
 }
 
+- (void)gotUserGetDetailData:(UserDetailModel *)udm error:(APIError *)err {
+	if (err == nil) {
+		NSDictionary *params = [[NSMutableDictionary alloc] initWithCapacity:1];
+		[params setValue:udm.gravatarHash forKey:@"gravatar_hash"];
+		[self setPrefs:params];
+	} else {
+		// Couldn't retrieve user data
+	}
+	[self setUserGravatarImage:udm.gravatarHash];
+	[self setUserLoggedInDetails:udm.username];
+}
+
+- (void)setUserGravatarImage:(NSString *)gravatarHash {
+	NSMutableString *gravatarURL = [[NSMutableString alloc] initWithString:@"http://www.gravatar.com/avatar/"];
+	if (gravatarHash != nil) {
+		[gravatarURL appendString:gravatarHash];
+	}
+	[gravatarURL appendFormat:@"?d=mm&s=%f", self.uiUserGravatar.frame.size.width];
+
+	self.uiUserGravatar.image = [UIImage imageWithData:[NSData dataWithContentsOfURL:[NSURL URLWithString:gravatarURL]]];
+}
+
+- (void)setUserLoggedInDetails:(NSString *)username {
+	NSString *loggedInText = [NSString stringWithFormat:@"Logged in as %@", username];
+	self.uiLoggedInText.text = loggedInText;
+}
+
+- (void)toggleSignedIn:(BOOL)userIsSignedIn {
```

Comment:
Please rename to `setSignedIn` or similar. toggle implies changing state based on current held internal state whereas this method requires the state to be passed in (which is much better!)